velox/functions/prestosql/WordStem.h

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <libstemmer.h>
+
+#include "velox/functions/Udf.h"
+#include "velox/functions/lib/string/StringImpl.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
+/// to get word stem
+class Stemmer {
+ private:
+  sb_stemmer* sbStemmer_;
+
+ public:
+  Stemmer(sb_stemmer* stemmer) : sbStemmer_(stemmer) {}
+  ~Stemmer() {
+    sb_stemmer_delete(sbStemmer_);
+  }
+
+  /// Get the word stem or NULL if out of memory
+  const char* stem(const std::string& input) {
+    return (const char*)(sb_stemmer_stem(
+        sbStemmer_,
+        reinterpret_cast<unsigned char const*>(input.c_str()),
+        input.length()));
+  }
+};
+} // namespace
+
+/// word_stem function
+/// word_stem(word) -> varchar
+///     return the stem of the word in the English language
+/// word_stem(word, lang) -> varchar
+///     return the stem of the word in the specificed language
+///
+/// It uses the snowball stemmer library to calculate the stem.
+/// https://snowballstem.org
+/// It provides Java implementation which is used in Presto as well
+/// as C implementation. Therefore, both Presto and Prestimissio
+/// would have the same word stem results.
+template <typename TExec>
+struct WordStemFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+ private:
+  thread_local static inline std::map<std::string, std::unique_ptr<Stemmer>>
```

Comment:
Please, remove thread_local and static.

Reply:
? but it's needed from a static method. without thread_local for now, it will shared by multiple threads without mutex.

Reply:
The method shouldn't be static either.

Reply:
Check out regex function's for an example. They have a similar cache of compiled regexes.

Reply:
I see. so switch to expression lifespan for the stemmers.

Reply:
one question,  different drivers have their own WordStemFunction instance, right? make sure I don't need mutex here.

Reply:
That's right. Each logical thread of execution, driver, will create its own instance of the function. Each function will run single-threaded (it may run on different threads, but not concurrently).